meta-client/src/main/java/org/apache/helix/metaclient/impl/zk/ZkMetaClient.java

```diff
@@ -266,18 +279,25 @@ public void asyncSet(String key, T data, int version, AsyncCallback.StatCallback
 
   @Override
   public void connect() {
-    // TODO: throws IllegalStateException when already connected
     try {
+      _zkClientConnectionMutex.lock();
       _zkClient.connect(_initConnectionTimeout, _zkClient);
+      // register this client as state change listener to react to ZkClient EXPIRED event.
+      // When ZkClient has expired connection to ZK, it sill auto reconnect until ZkClient
+      // is closed or connection re-established.
+      // We will need to close ZkClient when user set retry connection timeout.
+      _zkClient.subscribeStateChanges(this);
```

Comment:
Curious: Does this state change event going into the event queue with other type of events? Let's say if zk client connection is expired, other events (such as child change or data change) processing will fail because meta client cannot read zk now, will this block the queue so that zk state change event never gets to be processed? I saw that there is a method setAsyncExecPoolSize(), wondering is this how we prevent this kind of situation?

Reply:
TFTR. This is a good point.
As our offline discussed, since ZkClient has a single thread (_eventThread) handling all events, it is possible that one previous event handling callback blocking all following event being send out. This is a generic problem with single event thread. To coup with this, we
1. Mentioned explicitly in document that user should not register blocking callbacks to client. (ZooKeeper document also suggested this)
2. In Helix use case, all callbacks are just simply forward.